Fixes up the stress test CLI command

halgari opened the pull request:
Fixes the rather broken stress-test CLI command. Also fixes a lot of the flakey test issues we have in our test system. 

Previously: 

`IGame` would point to `IEnumerable<GameInstallation>` but `GameInstallation` also would point to `IGame`. In order to speed up the process of linking games to installations, `AGame` would cache Game installations. GameRegistry would be used by parts of the code to look up installations. Our manually added game locators would then also attempt to inject data into all these classes causing issues with caching. 

Now:

All these classes are condensed and their logic exists in `IGameRegistry`. Creating a new `GameInstallation` is as simple as registering it with `IGameRegistry` and then everything flows from there. `IGameRegistry` caches results and provides observable interfaces on all of these components. 


This means that in our tests we can say `StubbedGame.Create(...)` and get a brand-new stubbed game install, no clearing of caches required. Part of this PR also includes a `.zip` file that contains the contents of the `StubbedGame` install so that we have a ground-truth for all our game tests against the stubbed game. 

Fixing up the `stress-test` command required reworking some of the rendering code, so that's why the PR contains some of that data.

erri120 commented:
Do you also want to do #1414?

Sewer56 commented:
Misc comment: It's nice to see Skippable Facts not show as errors in the IDE

![image](https://github.com/Nexus-Mods/NexusMods.App/assets/6697380/f6bd38ba-5722-4b25-a22b-45dccfbbd683)

I remember we opted not to use them last year because they were getting reported as errors.

Sewer56 commented:
I'm not really sure of the culprit but when running the App from this PR, JSON serialization seems to have issues

First run:
```
00:00:00.054 [ERROR] Exception while serializing `NexusMods.DataModel.DataModelSettings` with value `DataModelSettings { UseInMemoryDataModel = False, MnemonicDBPath = XDG_DATA_HOME/NexusMods.App/DataModel/MnemonicDB.rocksdb, ArchiveLocations = NexusMods.Abstractions.Settings.ConfigurablePath[] }`|System.InvalidOperationException: ValueFactory attempted to access the Value property of this instance.
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at NexusMods.Abstractions.Settings.JsonStorageBackend.Serialize[T](Stream stream, T value) in /home/sewer/Project/NexusMods.App/src/Abstractions/NexusMods.Abstractions.Settings/JsonStorageBackend.cs:line 59
```

Subsequent Runs:
```
il: NexusMods.Abstractions.Settings.JsonStorageBackend[0]
      Exception while deserializing to `NexusMods.App.LoggingSettings` with JSON `{"MainProcessLogFilePath":{"BaseDirectory":"XDG_STATE_HOME","File":"NexusMods.App/Logs/nexusmods.app.main.current.log"},"MainProcessArchiveFilePath":{"BaseDirectory":"XDG_STATE_HOME","File":"NexusMods.App/Logs/nexusmods.app.main.{##}.log"},"SlimProcessLogFilePath":{"BaseDirectory":"XDG_STATE_HOME","File":"NexusMods.App/Logs/nexusmods.app.slim.current.log"},"SlimProcessArchiveFilePath":{"BaseDirectory":"XDG_STATE_HOME","File":"NexusMods.App/Logs/nexusmods.app.slim.{##}.log"},"MaxArchivedFiles":10,"MinimumLevel":"Debug","LogToConsole":true}`
      System.Text.Json.JsonException: The JSON value could not be converted to Microsoft.Extensions.Logging.LogLevel. Path: $.MinimumLevel | LineNumber: 0 | BytePositionInLine: 523.
         at System.Text.Json.ThrowHelper.ThrowJsonException(String message)
         at System.Text.Json.Serialization.Converters.EnumConverter`1.Read(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options)
         at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
         at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
         at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
         at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
         at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
         at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
         at NexusMods.Abstractions.Settings.JsonStorageBackend.Deserialize[T](String json) in /home/sewer/Project/NexusMods.App/src/Abstractions/NexusMods.Abstractions.Settings/JsonStorageBackend.cs:line 71
00:00:00.000 [DEBUG] Hosting starting
00:00:00.061 [ERROR] Exception while deserializing to `NexusMods.SingleProcess.CliSettings` with JSON `{"ListenTimeout":"00:00:01","StartCliBackend":true,"SyncFile":"/run/user/1000/NexusMods.App-sync_file.sync"}`|System.Text.Json.JsonException: The JSON value could not be converted to NexusMods.Paths.AbsolutePath. Path: $.SyncFile | LineNumber: 0 | BytePositionInLine: 107.
   at System.Text.Json.ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type propertyType)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
   at NexusMods.Abstractions.Settings.JsonStorageBackend.Deserialize[T](String json) in /home/sewer/Project/NexusMods.App/src/Abstractions/NexusMods.Abstractions.Settings/JsonStorageBackend.cs:line 71
00:00:00.074 [ERROR] Exception while deserializing to `NexusMods.DataModel.DataModelSettings` with JSON ``|System.Text.Json.JsonException: The input does not contain any JSON tokens. Expected the input to start with a valid JSON token, when isFinalBlock is true. Path: $ | LineNumber: 0 | BytePositionInLine: 0.
 ---> System.Text.Json.JsonReaderException: The input does not contain any JSON tokens. Expected the input to start with a valid JSON token, when isFinalBlock is true. LineNumber: 0 | BytePositionInLine: 0.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   --- End of inner exception stack trace ---
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, JsonReaderException ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo)
   at NexusMods.Abstractions.Settings.JsonStorageBackend.Deserialize[T](String json) in /home/sewer/Project/NexusMods.App/src/Abstractions/NexusMods.Abstractions.Settings/JsonStorageBackend.cs:line 71
00:00:00.076 [ERROR] Exception while serializing `NexusMods.DataModel.DataModelSettings` with value `DataModelSettings { UseInMemoryDataModel = False, MnemonicDBPath = XDG_DATA_HOME/NexusMods.App/DataModel/MnemonicDB.rocksdb, ArchiveLocations = NexusMods.Abstractions.Settings.ConfigurablePath[] }`|System.InvalidOperationException: ValueFactory attempted to access the Value property of this instance.
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at NexusMods.Abstractions.Settings.JsonStorageBackend.Serialize[T](Stream stream, T value) in /home/sewer/Project/NexusMods.App/src/Abstractions/NexusMods.Abstractions.Settings/JsonStorageBackend.cs:line 59
```

Seems to happen even after merging main and with a clean build.

Basically it seems `NexusMods.DataModel.DataModelSettings.json` fails to persist to disk, and on subsequent restarts of the App, fails to load, because an empty file was left in place.

I haven't tried to debug this yet, am just reviewing through the code.

Sewer56 commented on src/Abstractions/NexusMods.Abstractions.GameLocators/GameInstallation.cs:
```diff
@@ -74,4 +68,38 @@ public GamePath ToGamePath(AbsolutePath absolutePath)
     public IGameLocator Locator { get; init; } = null!;
+    
+    /// <summary>
+    /// An entity id that points to the game metadata in the mnemonidb instance
```
Got a typo there chief

Sewer56 commented on src/Abstractions/NexusMods.Abstractions.Cli/RendererExtensions.cs:
```diff
@@ -64,6 +64,7 @@ public static async Task<T> WithProgress<T>(this IRenderer renderer, Cancellatio
     public static async Task Error(this IRenderer renderer, Exception ex, string template, params object[] args)
     {
-        await renderer.Text(template + "\n\nTraceback: {ex} \n ", args);
+        await renderer.Text(template, args);
```
It's a cold path, so not really a big deal, but latency here would be lower if it was still 1 render message across IPC.

halgari replied:
I'm not really concerned as this is only used when throwing an exception, and the templating here was really strange otherwise, as we have to concat the arguments and the template

Sewer56 commented on src/Games/NexusMods.Games.AdvancedInstaller.UI/AdvancedManualInstallerUI.cs:
```diff
@@ -17,6 +17,13 @@ namespace NexusMods.Games.AdvancedInstaller.UI;
+    ///
+    /// NOTE(halgari): this is an ugly hack, but we don't have a better tool for it at the moment.
+    /// </summary>
+    public static bool Headless { get; set; } = false;
```
If this ever becomes a larger issue, then I think one possible pattern here is using DI.

We could use DI to inject an interface with 'debug' (or whatever you wanna call them) flags.

One of these flags could be `IsUIAvailable`, so if we're running unit tests, that could be used to enable headless mode as to not risk getting tests stuck.

This is just one possible approach that comes to mind.

halgari replied:
Yep, we can and should look into this.

Sewer56 reviewed (changes requested):
This fixes the Race Condition, thus I'm happy to close the temporary workaround on #1432

This also means the following hack in `uninstall-app` can now be removed:

```
// Prevent a race condition where `IRepository` is not fully done initializing.
// https://github.com/Nexus-Mods/NexusMods.App/issues/1396
Thread.Sleep(1000);
```

I left some comments/feedback above, but did not put them as part of the review because I'm not strictly requesting changes there.

In any case, I've got no issues with the code itself, however https://github.com/Nexus-Mods/NexusMods.App/pull/1419#issuecomment-2136201730 must be addressed before merging.

halgari commented:
The lazy serialization issue was because the datamodel needed settings which required the GameInstallation serializer, which required the GameRegistry which required the datamodel. Added another `Lazy` in there and that fixed it. Also simplified how we're saving GameInstallation values in settings. Now we use the MnemonicDB ID instead of tuples of metadata values. Having everything run off Ids is so much simpler.

Sewer56 commented:
(Just don't merge before @erri120 gets to it, as they still want to review)

Sewer56 commented:
Oh and don't forget to remove the hack in `uninstall-app` verb.

erri120 commented on src/NexusMods.SingleProcess/CliServer.cs:
```diff
@@ -137,6 +139,10 @@ private void CleanClosedConnections()
     /// <inheritdoc />
     public async Task StartAsync(CancellationToken cancellationToken)
     {
+        // Bit of a hack, but works for now till we get better startup logic with DI
```
Can you update the comment to explain why this is needed?

erri120 commented on src/NexusMods.DataModel/GameRegistry/Registry.cs:
```diff
@@ -19,110 +23,141 @@ namespace NexusMods.DataModel;
     {
-        await ((IHostedService)_conn).StartAsync(cancellationToken);
-        await Startup(_games);
+        lock (this)
+        {
+            _startupTask ??= Startup(cancellationToken);
+        }
+        await _startupTask;
```
Why are we locking here?

halgari replied:
There's a race condition here: 

C# example:

```csharp
using System;
public class C {
    int? i = null;
    public int M(int v) {
        i ??= v;
        return i.Value;
    }
}
```

compiles to

```csharp
using System;
using System.Diagnostics;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Security;
using System.Security.Permissions;

[assembly: CompilationRelaxations(8)]
[assembly: RuntimeCompatibility(WrapNonExceptionThrows = true)]
[assembly: Debuggable(DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints)]
[assembly: SecurityPermission(SecurityAction.RequestMinimum, SkipVerification = true)]
[assembly: AssemblyVersion("0.0.0.0")]
[module: UnverifiableCode]
[module: RefSafetyRules(11)]

public class C
{
    private Nullable<int> i;

    public int M(int v)
    {
        int valueOrDefault = i.GetValueOrDefault();
        if (!i.HasValue)
        {
            valueOrDefault = v;
            i = valueOrDefault;
        }
        return i.Value;
    }
}
```

So there's a small chance that wo threads may both try and start/set the value at the same time, the lock removes this chance.

erri120 commented on tests/NexusMods.CLI.Tests/VerbTests/apply_CanCreateAndManageLists.txt.verified.txt:
I think this file can be removed.

erri120 commented on tests/NexusMods.StandardGameLocators.TestHelpers/NexusMods.StandardGameLocators.TestHelpers.csproj:
```diff
@@ -12,5 +12,29 @@
         <None Update="xunit.runner.json">
           <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
         </None>
+        <None Update="Resources\StubbedGame.zip">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="Resources\StubbedGameState\game\Data\image.dds">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="Resources\StubbedGameState\game\config.ini">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="Resources\StubbedGameState\game\README.txt">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="Resources\StubbedGameState\game\StubbedGame.exe">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="Resources\StubbedGameState\preferences\settings.ini">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="Resources\StubbedGameState\saves\save1.dat">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
+        <None Update="Resources\StubbedGameState\saves\save2.dat">
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+        </None>
```
Should be cleaned up to copy the folder instead of the files.

halgari commented:
Ended up having to go back to the files being packed in the .zip file. Something about git/windows and a specific ini file meant the hashes that weren't matching. They'd match on my machine but something along the way would mangle them and add character to the files. Using a .zip means that git and/or windows won't try to muck with text files.
